src/dpq.rs

```diff
@@ -31,3 +34,91 @@ pub fn r_dt_1(lower_tail: bool, log_p: bool) -> f64 {
         r_d__0(log_p)
     }
 }
+
+pub fn r_d_lval(p: f64, lower_tail: bool) -> f64 {
+    if lower_tail {
+        p
+    } else {
+        // Using 0.5 - p + 0.5 to perhaps gain 1 bit of accuracy.
+        0.5 - p + 0.5
+    }
+}
+
+pub fn r_d_cval(p: f64, lower_tail: bool) -> f64 {
+    if lower_tail {
+        // Using 0.5 - p + 0.5 to perhaps gain 1 bit of accuracy.
+        0.5 - p + 0.5
+    } else {
+        p
+    }
+}
+
+pub fn r_dt_qiv(p: f64, lower_tail: bool, log_p: bool) -> f64 {
+    if log_p {
+        if lower_tail {
+            p.exp()
+        } else {
+            -expm1(p)
+        }
+    } else {
+        r_d_lval(p, lower_tail)
+    }
+}
+
+pub fn r_dt_civ(p: f64, lower_tail: bool, log_p: bool) -> f64 {
+    if log_p {
+        if lower_tail {
+            -expm1(p)
+        } else {
+            p.exp()
+        }
+    } else {
+        r_d_cval(p, !lower_tail)
+    }
+}
+
+/// Calculate the boundaries exactly for q*() functions.
+/// Often left = ML_NEGINF, and very often right = ML_POSINF;
+///
+/// R_Q_P01_boundaries(p, left, right)  :<==>
+///
+/// R_Q_P01_check(p);
+/// if (p == R_DT_0) return left;
+/// if (p == R_DT_1) return right;
+///
+/// the following implementation should be more efficient (less tests):
+///
+/// This was originally a macro, but it is now a function.
+/// At the caller site, if the return value is not None, then return the
+/// result immediately.
```

Comment:
Pay attention with macro's that return. When converting to a function, make sure to propagate the immediate return upwards. I did it here with an `Option`. @storopoli agreed?